src/ripple/app/tx/impl/apply.cpp

```diff
@@ -86,15 +93,79 @@ invoke_apply (ApplyContext& ctx)
 
 //------------------------------------------------------------------------------
 
+std::pair<Validity, std::string>
+checkValidity(HashRouter& router, STTx const& tx, bool allowMultiSign)
+{
+    auto const id = tx.getTransactionID();
+    auto const flags = router.getFlags(id);
+    if (flags & SF_SIGBAD)
+        // Signature is known bad
+        return std::make_pair(Validity::SigBad,
+            "Transaction has bad signature.");
+    if (!(flags & SF_SIGGOOD))
+    {
+        // Don't know signature state. Check it.
+        if (!tx.checkSign(allowMultiSign))
```

Comment:
`checkSign` could possibly be rewritten as a free function. `STTx` doesn't have many private members. That said, I think it should stay, partially because it fits logically, but also because my understanding of the SusPay hash/signature/proof/digest check could be _moved_ from SusPay.cpp to `STTx`. Obviously, I could be wrong, but a quick look at the three relevant fields: `sfMethod, sfDigest, sfProof` are unique to SusPays, and if so, then any misuse of them in another transaction could be treated as a signature failure.